firmware-management/mbl-watchdog-init/mbl-watchdog-init.c

```diff
@@ -0,0 +1,173 @@
+/*
+ * Copyright (c) 2019 Arm Limited and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ *
+ * This utility initialises the hardware watchdog and sets the timeout.
+ * The program also tries to determine the last boot reason so we can tell if
+ * the last reboot was caused by the hardware watchdog.
+ * This code is intended for use with our robust update mechanism; we need
+ * to start the watchdog before we mount the rootfs, so the device reboots if
+ * loading the rootfs fails.
+ */
+
+#include <errno.h>
+#include <fcntl.h>
+#include <getopt.h>
+#include <linux/watchdog.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/ioctl.h>
+#include <unistd.h>
+
+//============================================================================
+// Error handling and logging convenience functions
+int is_err(const int ret)
+{
+    return ret == -1;
+}
+
+
+int log_error(const char *message)
+{
+    int err_code = errno;
+    char *prefix = "WATCHDOG-ERROR:";
+    char result[strlen(prefix) + strlen(message)];
```

Comment:
I'm going to change this to a fixed length array, but keep it on the stack. What size would you recommend so we don't go around this loop again?

Reply:
After reading https://stackoverflow.com/questions/36318928/c-c-maximum-size-of-errno-associated-strings-at-compile-time, I don't think you can deduce a reasonable fixed buffer size that's guaranteed to work. I recommend putting the buffer on the heap.

https://stackoverflow.com/questions/36318928/c-c-maximum-size-of-errno-associated-strings-at-compile-time also suggests using `strerror_r` rather than `strerror`, which sounds like a good idea to me.

Reply:
Next commit changes this so it just uses `fprintf` rather than allocating space for the string buffer or using a variable length array. Let me know what you think.

Reply:
Ah, yeah, that sounds much better... Now that you've said it it seems obvious ;-)

Reply:
f631d22ca958a9b52e3ca034b8ba5199d0d1117a